docs/cip/CIP-10052023_Estimation_Tool.md

```diff
@@ -0,0 +1,150 @@
+- 🔥 Load Factor - a multiplier used to model the anticipated load on Chroma, this used to model variations in query complexity, data volume, temporary objects and partly concurrency. We suggest four simplified load factors:
+  - Low - 1.0 (default value)
+  - Medium - 1.2
+  - High - 1.5
+  - Burst - 2.0
+- ⧥ Concurrency Factor - a multiplier used to model anticipated parallel queries in Chroma. The intuition is that as the higher the number of concurrent queries, the higher the number of temporary objects used by Chroma. This metric is mostly non-linear therefore we suggest four simplified concurrency factors:
+  - Low - 1.0 (default value)
+  - Medium - 1.2
+  - High - 1.5
+  - Burst - 2.0
+- 🖥️ System Overhead - a multiplier used to model the overhead of the system (e.g. OS, Chroma server, etc.). We suggest a 30% overhead for the system.
+
+Assumptions:
+
+- We assume that the user will rely on default hnsw hyper parameter configuration (`hnsw:space=l2`, `hnsw:construction_ef=100`, `hnsw:search_ef=10`, `"hnsw:M=16`) (Note: in a future iteration we can add a way to specify custom hnsw hyper parameters to the estimate)
+- We also assume default hnsw index batch parameters (`hnsw:batch_size=100`, `hnsw:sync_threshold=1000`)
+
+The final formula is as follows:
+
+`memory required = Vc * Vd * 4 bytes * Lf * Cf * Of`
+
+Where:
+
+- `Vc` - 💯 number of vectors
+- `Vd` - ⒩ vector dimensionality
+- `Lf` - 🔥 load factor
+- `Cf` - ⧥ concurrency factor
+- `Of` - 🖥️ system overhead factor
+
+The required memory is then rounded up to the nearest hardware increment.
+
+We suggest the following hardware increments:
+
+- 1 vCPU / 0.5 GB RAM
+- 1 vCPU / 1 GB RAM
+- 1 vCPU / 2 GB RAM
+- 2 vCPU / 4 GB RAM
+- 4 vCPU / 8 GB RAM
+- 8 vCPU / 16 GB RAM
+- 16 vCPU / 32 GB RAM
+- 32 vCPU / 64 GB RAM
+- 64 vCPU / 128 GB RAM
+- 128 vCPU / 256 GB RAM
+
+> Note: The above list is not an exhaustive one but can serve as basis for calculation.
+
+### Intuitions
+
+Our base intuition for the estimation is that generally memory is more important than CPU for most Chroma Workloads, however as things with load and concurrency scale CPU becomes more important.
+Another intuition we have is that disk IO may also be a bottleneck for some workloads, e.g. write heavy, but also some read heavy workloads such as multiple concurrent queries.
+Intuition for the growth factor is that as most useful systems are rarely single-user or static and as such they grow in at least two dimensions - the data stored and the load on the system. Our core intuition about this is that data growth is a function of user growth which in turn models the load on the system.
+
+### Future Work
+
+The current proposal is for a single instance Chroma deployment. However in the future we'd want this to work for distributed Chroma where more factors are at play.
```

Comment:
How could we make the cli tool aware of what backend is running?

Reply:
Version string or pre-flight can be a way to detect whether we're running in distributed or single node.